#162163267 create view and edit user profile

SokoPaulSokool opened the pull request:
### what does this PR do?
Adds create view and edit of user profile

## Description of tasks to be completed
writing unit tests for this feature
Have these endpoints working
POST /api/user
PUT /api/user
GET /api/profiles/:username

## How this can be tested manually

Clone the repo,
 Install the requirements with pip install -r requirements.txt
To run tests, run ./manage.py test

Run python manage.py runserver
Test the endpoints in postman

### Profile creation
Profile is created when user signs up

## To test this
Send post request to http://127.0.0.1:8000/api/users/
with body
```
{
"user":{
	"email":"soko@gmail.com",
	 "username":"soko",
	"password":"sokosoko"
	}
}
```
Then login that user by sending  a POST request to http://127.0.0.1:8000/api/users/login/
Login data format --> POST /api/users/login/
with body
```
{
  "user":{
	"email":"soko@gmail.com",
	"username":"soko",
	"password":"sokosoko"
}
 }
```
get the token returned

then send a GET request to http://127.0.0.1:8000/api/profiles/soko
![image](https://user-images.githubusercontent.com/10868919/49646246-9c732980-fa2f-11e8-852f-0d7ba487a1ce.png)

## Test update profile
Send a PUT request to http://127.0.0.1:8000/api/user/
with body and authentication token
```
{
  "user":{
    "email": "paul@andela.com",
    "bio": "I like to skateboard",
    "image": "https://www.andela.com/paul.jpg"
  }
}
```

Commit:
feat(profile): add files for profile app

Add all the initial files required for the profile application

[Delivers #162163267]

Commit:
feat(profile): add files for profile app

Add all the initial files required for the profile application

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add retrive profile functionlity

This adds routes to connect the profile app. Define model for Profile and a relationship to User. Create view to retrive profile

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add retrive profile functionlity

This adds routes to connect the profile app. Define model for Profile and a relationship to User. Create view to retrive profile

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add renderer class to format Profile

This commit adds a ProfileJSONRenderer class to renderers.py file. The class when added to renderer_classes in the views formarts the returned profile and lets the default renderer handle exceptions raised

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add renderer class to format Profile

This commit adds a ProfileJSONRenderer class to renderers.py file. The class when added to renderer_classes in the views formarts the returned profile and lets the default renderer handle exceptions raised

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add signal to create profile on user signup

This commit adds a signal that is run when the user is created the first time and. This signal the creation of a profile for that user.

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add signal to create profile on user signup

This commit adds a signal that is run when the user is created the first time and. This signal the creation of a profile for that user.

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add signal to create profile on user signup

This commit adds a signal that is run when the user is created the first time and. This signal the creation of a profile for that user.

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add signals file

This commit adds signals file with the create profile function

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add update function to profile

This commit adds an update function to the ProfileRetrieveView

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add update function to profile

This commit adds an update function to the ProfileRetrieveView

[Delivers #162163267]

SokoPaulSokool committed:
(feat): add require authentication to profile endpoints

This commit adds authentication to the profile endpoints. This makes them requie jwt token to be accessed

[Delivers #162163267]

SokoPaulSokool committed:
(feat): add require authentication to profile endpoints

This commit adds authentication to the profile endpoints. This makes them requie jwt token to be accessed

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add test for geting user profile

This commit adds tests for getting a profile of an existing user

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add test for geting user profile

This commit adds tests for getting a profile of an existing user

[Delivers #162163267]

collin5 reviewed (changes requested):
Have you fixed issues with edit `bio` and `profile_pic`? Also add tests for this.

KicaRonaldOkello committed:
Feature(Documentation):Documentation using swagger (#11)

joshtrigger committed:
#162163265 send user an email with link to reset password (#13)

SokoPaulSokool committed:
feat(profile) add edit bio and image functionality

This commit adds edit image and bio functionality when a user sends a post request

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile) add edit bio and image functionality

This commit adds edit image and bio functionality when a user sends a post request

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add edit bio and image functionality

This commit adds a functionality to edit bio and image when user sends PUT request to /api/user/
[Delivers #162163267]

SokoPaulSokool committed:
feat(profile): add test for update bio and image

This commit adds test for update bio and image

[Delivers #162163267]

SokoPaulSokool committed:
feat(profile) add check for getting profile if it exists

This commit adds a check if the profile one is trying to get exists. Adds ProfileDoesNotExist class to raise exception if it does not exist. It also removes update function from ProfileRetrieveView because it was not being used anywhere. Also adds test for this check

[Delivers #162163267]

SokoPaulSokool commented:
I have fixed the issues with edit bio and image and added the tests
